Complete rewrite of DFRN generation

annando opened the pull request:
The generation of the XML for the DFRN protocol is completely rewritten. It now uses the XML DOM functions and is split into several sub functions into a new file. Additionally the old code was removed.

The protocol elements have been reviewed. Elements that aren't necessary (e.g. when they belonged to the OStatus protocol) are removed. Several elements are new. In the header we now transmit data from the default user profile. (OStatus is doing it as well) Some additional fields are now transmitted (like the keywords or the timezone - but only in the direct server-server communication). A preparation was done for XMPP.

It now worked stable on my system for some time. The elements were checked for completeness. Only the relocation wasn't tested "in the wild" but only with test functions.

rabuzarus commented on include/dfrn.php:
```diff
@@ -0,0 +1,791 @@
+		return $author;
+
+	// Only show contact details when we are allowed to
+	$r = q("SELECT `profile`.`about`, `profile`.`name`, `profile`.`homepage`, `user`.`nickname`, `user`.`timezone`,
```
This query looks like something, which could be also useful at other places and should maybe a separate function

annando replied:
Do you mean just the line 468? Or do you mean the whole "poco" section? I don't think that we could use it in another place. But of course I could separate it in that file. (Just to make the code clearer)

rabuzarus replied:
I mean getting all profile fields for a given uid.

annando replied:
I don't agree with that. If we always would use abstract calls instead of direct queries then I would agree. We don't even have some code logic in that query (in difference to the "get_contact_details_by_url" function). And that query would have to include many different fields from three different tables (profile, user, contact) to be complete. And last but not least I don't see many places were we would need that call.

rabuzarus commented on include/Contact.php:
```diff
@@ -662,4 +664,34 @@ function posts_from_contact($a, $contact_id) {
+ *
+ * @return string Location string
+ */
+function formatted_location($profile) {
```
I'm not sure if we should have this function in include/Contacts.php   There are already some some format_* functions in include/text.php. Even it is a very large file it should be placed there for consistency reasons. What are you thinking?

annando replied:
I'm not sure about that. We really have to move many functions to better places. I would like to split the large files into many small files as well. I thought the "Contact.php" would be the right place because the function is used for contact data.

rabuzarus replied:
Mhhh I'm also not really sure. text.php is really bloated. On the other formatted_location doesn't seem to fit totally because it doesn't do anything with the contact and it is more related to profile or identity things. But I don't have a real good  alternative.

annando replied:
You don't have time in the next days? I'm having holidays. We could do a "mini hackathon" where we could discuss stuff like this.

rabuzarus replied:
I have to write a test in on Wednesday. It is exams time at the end of the semester :-( Because I have no real good organization I don't really know how much time I will have in the next weeks. Maybe I have some time, maybe not :-/ 
But I have to say that I'm really interested in such things.

BTW: I wan't to give you respect for the large work on code cleaning. It's really awesome.

Ahh and if I find some time I will try to finish https://github.com/rabuzarus/friendica/tree/1901_autocomplete (I don't know if it is wanted. Originally it should be only a test. To sum up: It's a port of hubzillas autocomplete feature with the possibility to have autocomplete in the nav search for all gcontact contact entries)

annando commented:
@rabuzarus Would you see any of the mentioned points as a blocker?

rabuzarus commented:
No no, this points were only meant as suggestions.

annando commented:
I hope that my arguments don't sound like from an ignorant person? BTW: Is the documentation alright?

rabuzarus commented on include/dfrn.php:
```diff
@@ -0,0 +1,791 @@
+ * @param string $dfrn_id
+ * @param string $owner_nick Owner nick name
+ * @param string $last_update Date of the last update
+ * @param int $direction
```
What does direction mean? What values can be set? 0 and 1?

annando replied:
Direction can be -1, 0 and 1 - and I really don't know what this value is for.

rabuzarus replied:
OK this happens :-D Then add something like `Available values can be -1, 0 or 1`

rabuzarus commented on include/dfrn.php:
```diff
@@ -0,0 +1,791 @@
+ * @param array $item Item element
+ * @param array $owner Owner record
+ * @param bool $comment Trigger the sending of the "comment" element
+ * @param int $cid
```
- The id of the contact

annando replied:
I guess so. It is used for the private photos - but it is part of the structure that I reused.

rabuzarus commented:
@annando 
No they don't sound ignorant. I wasn't about this points I mentioned above but I thought it is better to say/ask this things because if they are valid points it could help and if they don't make much sense then somebody can tell me this. If something like this becomes annoying in future you have to tell me this.

According to the docu. It is looking fine. At some point there could be some more inline docu. Not because nobody would understand what happens but more to the fact to understand why this have to happen (since it is the protocol and so one of the most important parts of friendica it would explain more how friendica is working).

And I added 2 minor points (look above) where it lacks in description (but they are really marginl ;-)  )

annando commented:
I don't know the protocol in completeness either. I guess I will get to know it better when I'm rewriting the receiver. There are some fields where I really don't know what they are for.

annando commented:
And some more documentation :-)

rabuzarus commented:
:+1: 

Sorry for pointing every time to this topic. To write docu is very annoying and boring. But it helps other and in this sense it helps 'us' too.

After overflowing the code 4 times it's more and more fun to read in contrast to the old code

rabuzarus commented:
Just looking at your last commit. Yeah something like this is really wonderful ;-)

annando commented:
BTW: You carry out a very important function by pointing your (virtual) finger at things like this. Thanks a lot for this!

rabuzarus commented on include/dfrn.php:
```diff
@@ -0,0 +1,795 @@
+	$converse    = false;
+
+	if($public_feed && $a->argc > 2) {
+		for($x = 2; $x < $a->argc; $x++) {
```
BTW: we now have the functions argc()  (https://github.com/friendica/friendica/blob/develop/boot.php#L2095) and argv($x) (https://github.com/friendica/friendica/blob/develop/boot.php#L2105) so you don't need to commit the whole app $a to a function anymore to get argv or argc

rabuzarus commented on include/dfrn.php:
```diff
@@ -0,0 +1,795 @@
+	$photos = array();
+	$ext = Photo::supportedTypes();
+	foreach($rp as $p){
+		$photos[$p['scale']] = $a->get_baseurl().'/photo/'.$p['resource-id'].'-'.$p['scale'].'.'.$ext[$p['type']];
```
you can use `z_root()` instead of `$a-get_baseurl()` so you wouldn't need to load the whole app `$a = get_app()`

rabuzarus commented on include/dfrn.php:
```diff
@@ -0,0 +1,795 @@
+ * @return object XML author object
+ */
+function dfrn_add_entry_author($doc, $element, $contact_url, $item) {
+	$a = get_app();
```
is the app needed for this function?

annando replied:
No. This was left over when I moved some code.

rabuzarus commented:
Thank you for the flowers.

I read the code again. There are many places where you load the app (`$a = get_app();`) just to get the baseurl. You can also use the function `z_root()` for this. I didn't mark all the corresponding lines because I don't know if you want to.

annando commented:
I must confess that I'm not a big fan of using the functions "argc", "argv" or "z_root". In the background they are doing the same like $a->argc or $a->rgv would do. So they don't help with the execution time or the memory foot print. They aren't a benefit in readability as well, I guess.

While writing these lines I had the thought that we should use static class calls and singletons for our new code. I guess I will rewrite the DFRN code and the OStatus code according to this. This would help to increase readability since we would only expose a few functions but the rest would be private.

I think the same should be valid for functions like z_root. If we would replace them with static class calls like "app::root" this would increase readability (since we would always know where this function comes from) and internally it could use the necessary variables.

What do you think?

rabuzarus commented:
:+1:

annando commented:
I now changed the code so that "dfrn" is a class. Since the methods are stateless they can be called statically without a problem. Additionally I changed "get_baseurl" so that it can be called statically as well.

rabuzarus commented on include/dfrn.php:
```diff
@@ -0,0 +1,1024 @@
+<?php
+require_once('include/items.php');
```
```
/**
 * @file include/dfrn.php
 * @brief The implementation of the dfrn protocol
 *
 * https://github.com/friendica/friendica/wiki/Protocol
 */

rabuzarus commented on include/dfrn.php:
```diff
@@ -0,0 +1,1024 @@
+<?php
```
```
/**
 * @file include/dfrn.php
 * @brief The implementation of the dfrn protocol
 *
 * https://github.com/friendica/friendica/wiki/Protocol
 */
```

rabuzarus commented on include/dfrn.php:
```diff
@@ -0,0 +1,1024 @@
+require_once('include/items.php');
+require_once('include/Contact.php');
+require_once('include/ostatus.php');
+
```
the dfrn class needs at least a `@brief` description

rabuzarus commented on include/dfrn.php:
```diff
@@ -0,0 +1,1024 @@
+require_once('include/Contact.php');
+require_once('include/ostatus.php');
+
+class dfrn {
```
the dfrn class needs at least a `@brief` description

annando commented:
And still some more documentation :-)

annando commented:
@rabuzarus Do I have your go?

rabuzarus commented:
@annando You have it :-P
